### DISCUSSION TITLE
Rules for username.

### DISCUSSION THREAD

**Discussion by yasukawa426** [zitadel/zitadel#3956] [Ideas]
Right now you are able to write anything you want in your username, for example, you can use whitespace or even emojis. It would be a good idea to be able to create rules to username just like you can with the passwords. 

![image](https://user-images.githubusercontent.com/62857460/178745633-7afb638e-2107-45a3-8903-39847ae7079f.png)

**Comment by fforootd** [comment]
Hm, I find that an Interesting idea.
@hifabienne what do you think of this?

**Reply by hifabienne** [reply]
Yes I think this is something that could be interesting for different customers. I could also imagine that you like to create some kind of rules how usernames will be created. (e.g firstname or firstname.lastname)

Its not something we have currently on our roadmap, but if enough customers would like to have this we will have a look about the implementation. 

Otherwise we are also open for PRs. 😉

**Reply by mffap** [reply]
Somewhat related: We should consider trimming the white space around the loginname in the hosted login page. Some autofill actions seem to like putting a whitespace at the end. This would only work consistently when a user can disallow whitespaces.

**Comment by shedar** [comment]
I have a related question/suggestion. For now, the username field is mandatory and used for login. We'd prefer using emails for signing in. As a workaround, we can set the username to the same value as the email, but it can be confusing for users and looks unexpected in some places (e.g., account selection shows "john@gmail.com@organization" even with org suffixes turned off in settings).
Is there a way/plan to allow using just emails?

**Reply by fforootd** [reply]
Hi @shedar 

Great question.

You can choose to ignore the org suffix on an organization by disabling the `User Loginname must contain orgdomain` under an orgs settings.

![image](https://user-images.githubusercontent.com/9879976/179497806-b747bb4d-1d3b-474c-80b6-947ec559b4e5.png)

For "Business Customers" in B2B scenarios they have[ the option to self-service verify their domain](https://docs.zitadel.com/docs/guides/basics/organizations#exercise---verify-your-domain-name) and use it as orgname. Hence allowing them to use their suffix.

Relates to [#3967
](https://github.com/zitadel/zitadel/discussions/3967)

**Reply by shedar** [reply]
Hi @fforootd,

Thank you for the suggestions. It would definitely make this approach more viable. But from user management/profile editing experience, it still would be error-prone, e.g., when someone changes the email address but forgets to change the username unless there's to automate syncing the fields (event hooks or something similar)

**Reply by fforootd** [reply]
Interesting thought. What we often see is that companies who use zitadel changes such things through our api

So for example if a user changes their email in the customers portal, they write to our apis twice. Once to update the mail and once the username.

@hifabienne I think we should think about an potential improvement here. What do you think of something like a "synced" or "mail as login" mode on an org?

**Reply by mffap** [reply]
Why not handle the sync with an Action? I know that hook is not available, yet, but could be useful for a couple of things when a user updates their mail address.

**Reply by fforootd** [reply]
True, that could work as well 😁

It would need two triggers (which we still need to enable) to the action. One if the username changes and one if the mail changes.
But the same code could be used :-)

**Reply by hifabienne** [reply]
I thing this could be a usecase a lot of custoemers could have. One thing we would need to look at is the validation of the email address, at the moment the usernames doesn't have a verification, but the email does.

**Reply by danielloader** [reply]
@hifabienne was there ever further discussion on using emails as usernames? I'd rather not store/use both fields for a project I'm working on.

**Reply by hifabienne** [reply]
We do want to provide the option of user schemas in combination witch custom actions. this would mean that you can add a custom  action which for example syncs the email with the username if needed.

**Comment by fforootd** [comment]
Since this came up more and more I am going to convert this to an actual issue.
The latest trigger is this issue here https://github.com/zitadel/zitadel/issues/4245
